### PR TITLE
fix #218 by displaying sound durations in HH:MM:SS format

### LINKED ISSUE #218: bug: sounds have SS:XX format instead of MM:SS format
Sounds like https://freesound-player.vercel.app/sound/263662 that are over an hour in length shouldn't display the minutes section of their length with a value over 59.

Screen Shot 2020-11-21 at 2 19 48 PM

### PR DESCRIPTION
**Pull request by hamirmahal**
This pull request fixes #218. Sounds in the Pack and Similar `<SoundList />` components had hitherto displayed durations in a SS:XX format. This pull request should make it so that those durations now display in MM:SS format for sounds less than an hour long, and HH:MM:SS format for sounds over an hour in duration.

I made two other minor changes.

Firstly, sounds over an hour in length needed more room to display their duration, so I changed the percentage width of sound durations from ~33% to 50%.

Secondly, I noticed an inconsistency between the stated length of sounds on our application and sounds downloaded to a device.

For instance, downloading one track that our application listed as 8 seconds long actually appeared as 9 seconds long in Finder after downloading it.

So, I changed our approach from flooring the incoming sound duration to rounding it:﻿this seems to create a consistent sound duration across our application and the stated duration of downloaded files.

### CODE REVIEW

**Comment by hamirmahal** [comment]
To get an idea of the kind of changes this pull request makes, we can look at a sound with a really long length, e.g., https://freesound-player.vercel.app/sound/376019, and compare that to https://freesound-player-git-fix-sound-length-formatting.amilajack.vercel.app/sound/376019.

Current appearance of our application at the time of this writing

Screen Shot 2020-11-26 at 1 25 27 PM

Appearance with the changes in this pull request

Screen Shot 2020-11-26 at 1 26 26 PM

File: src/SoundList.js [hunk 1]
```diff
@@ -3,10 +3,15 @@ import { Link } from "react-router-dom";
 import { playCircleSharp, timeSharp, downloadSharp } from "ionicons/icons";
 import { IonIcon } from "@ionic/react";
 
-function timeConvert(minutes) {
-  const min = Math.floor(Math.abs(minutes));
-  const secs = Math.floor((Math.abs(minutes) * 60) % 60);
-  return `${(min < 10 ? "0" : "") + min}:${secs < 10 ? "0" : ""}${secs}`;
+function format(lengthInSeconds) {
```
Comment (amilajack) [inline src/SoundList.js:6]: This algorithm seems more complicated than the previous one. I'm imagining that the solution could be a simple addition to the existing algorithm.

I'm imagining the structure could look like this:
```
const secs = ...
const min = ...
const hrs = ...
```

Also I think `minutes` is a better name for a param since it isn't coupled to the context of our app. In other contexts, "length" in minutes might not make sense. The goal for our utilities/components should be to have the minimal amount of coupling to our app.

**Review by amilajack — CHANGES_REQUESTED** [review]

File: src/SoundList.js (hunk 1 above)
Comment (amilajack) [reply src/SoundList.js:6]: I also think `format` isn't a good function name. Readers think "format what"? I think `formatTime` is more clear. The variable names `m` and `h` are also an antipattern. Our convention is to never use single letter variable names. `min`, `hrs`, and `secs` are more clear.

Comment (hamirmahal) [reply src/SoundList.js:6]: > I also think `format` isn't a good function name. Readers think "format what"? I think `formatTime` is more clear. The variable names `m` and `h` are also an antipattern. Our convention is to never use single letter variable names. `min`, `hrs`, and `secs` are more clear.

I used `format` to make it more declarative. I think `format(time)` is more readable than `formatTime(time)`. As far as the variable names go, I can definitely make them longer. They were verbose to begin with, but I shortened them because `prettier` wanted them to be on separate lines. I was trying to reduce the LOC to minimize vertical scrolling, since I read that having to scroll vertically, excessively, reduces readability.

> This algorithm seems more complicated than the previous one. I'm imagining that the solution could be a simple addition to the existing algorithm.
> 
> I'm imagining the structure could look like this:
> 
> ```
> const secs = ...
> const min = ...
> const hrs = ...
> ```
> 
> Also I think `minutes` is a better name for a param since it isn't coupled to the context of our app. In other contexts, "length" in minutes might not make sense. The goal for our utilities/components should be to have the minimal amount of coupling to our app.

I was trying to speed up computation time by having the function `format` compute as little as necessary. If the duration is less than a minute long, we don't even need to worry about formatting the minutes or hours portion of the time.

We should leave `seconds` in the parameter name: [according to the Freesound API Documentation](https://freesound.org/docs/api/resources_apiv2.html#response-sound-instance), `track.duration`, which is the only input we give this function, has a unit of seconds. We can change the parameter title to `durationInSeconds` instead of `lengthInSeconds`. It's what I had initially, but I was again trying to cut back on vertical scrolling, since the longer parameter name caused `prettier` to want to split `if (lengthInSeconds < 60) return `00:${addZeroToFrontOf(lengthInSeconds)}`;` across two lines instead of one.

I, for sure, have to make at least one more change, since pages like [the homepage](https://freesound-player-git-fix-sound-length-formatting.amilajack.vercel.app) look awkward with so much space given to the time for its formatting. It looks okay if the `<SoundList />` has a track over an hour long. If all sounds in a `<SoundList />` are under an hour long, like on the linked page, it doesn't look elegant. I have a fix for that.

Comment (amilajack) [reply src/SoundList.js:6]: > I used format to make it more declarative. I think format(time) is more readable than formatTime(time)...

Editors and other tools don't provide autocompletion based on variable names. They're always based on function name:

Screen Shot 2020-11-26 at 7 51 06 PM

When a reader is trying to search fns to get a list of fn's available in the current module, they'll see `format`, which won't make much sense to then since they can't see the params.

TL;DR: params should not be used to infer what a function does. Function name + comments take that responsibility.

> I was trying to speed up computation time by having the function...

This is a subtle form of premature optimization. The formatting of time is far from being the bottleneck of our application so sacrificing simplicity for complexity won't help us much.

> They were verbose to begin with, but I shortened them because prettier wanted them to be on separate lines...

I recommend not changing the code to reflect make it easier for prettier. I used to do this but it always bit me later when trying to understand my code later on. Single variable names are always an antipattern. Adding a few lines of code is fine.

See this section of [Idomatic.js](https://github.com/rwaldron/idiomatic.js/), which I highly recommend reading:
Screen Shot 2020-11-26 at 7 49 29 PM

Comment (hamirmahal) [reply src/SoundList.js:6]: > Editors and other tools don't provide autocompletion based on variable names. They're always based on function name:

I didn't realize editors omitted parameter names when looking up function names. Now I see that it does make more sense to add more detail to a function's name, like you said. My incoming commit will do that. Thanks for the tip!

> I recommend not changing the code to reflect make it easier for prettier. I used to do this but it always bit me later when trying to understand my code later on. Single variable names are always an antipattern. Adding a few lines of code is fine.

For sure. My incoming commit will have a lot more detail. I perused the [Idomatic.js](https://github.com/rwaldron/idiomatic.js/) link you sent me a bit last night, and it seems really useful. Thanks for sending me the link!

**Comment by hamirmahal** [comment]
By the way, certain really short sounds display a duration of `00:00`. While I initially thought this looked odd, it actually matches the duration that appears when downloading a sound with that duration and viewing its metadata or playing it in a media player, at least when I tried previewing it in Finder and playing it in QuickTime Player on macOS Big Sur, so I decided to leave that as is.

Another thing that this commit does is display all sounds in the HH:MM:SS format, instead of just those sounds over an hour in length. I initially thought this looked odd, too, since a ton of sounds less than a minute in length had a lot of zeros, but, on the other hand, it creates a consistent format, HH:MM:SS, across all sound pages, so I figured I would let that be, as well. Do you think these two changes are fine?

Current appearance of this application without the changes from this pull request on a sound page with a really long length:

Screen Shot 2020-11-27 at 10 32 49 AM

Appearance of this application with the changes from this pull request on a sound page with a really long length:

Screen Shot 2020-11-27 at 10 34 08 AM

Current appearance of this application's homepage without the changes from this pull request:

Screen Shot 2020-11-27 at 10 37 36 AM

Appearance of this application's homepage with the changes from this pull request:

Screen Shot 2020-11-27 at 10 36 23 AM

**Comment by hamirmahal** [comment]
By the way, certain really short sounds display a duration of `00:00:00`. While I initially thought this looked odd, it actually matches the duration that appears when downloading a sound with that duration and viewing its metadata or playing it in a media player, at least when I tried previewing it in Finder and playing it in QuickTime Player on macOS Big Sur, so I decided to leave that as is.

Another thing that this commit does is display all sounds in the HH:MM:SS format, instead of just those sounds over an hour in length. I initially thought this looked odd, too, since a ton of sounds less than a minute in length had a lot of zeros, but, on the other hand, it creates a consistent format, HH:MM:SS, across all sound pages, so I figured I would let that be, as well. Do you think these two changes are fine?

Current appearance of this application without the changes from this pull request on a sound page with a really long length:

Screen Shot 2020-11-27 at 10 32 49 AM

Appearance of this application with the changes from this pull request on a sound page with a really long length:

Screen Shot 2020-11-27 at 10 34 08 AM

Current appearance of this application's homepage without the changes from this pull request:

Screen Shot 2020-11-27 at 10 37 36 AM

Appearance of this application's homepage with the changes from this pull request:

Screen Shot 2020-11-27 at 10 36 23 AM

**Review by hamirmahal — COMMENTED** [review]
What other changes should I make?

File: src/SoundList.js [hunk 2]
```diff
@@ -3,10 +3,24 @@ import { Link } from "react-router-dom";
 import { playCircleSharp, timeSharp, downloadSharp } from "ionicons/icons";
 import { IonIcon } from "@ionic/react";
 
-function timeConvert(minutes) {
-  const min = Math.floor(Math.abs(minutes));
-  const secs = Math.floor((Math.abs(minutes) * 60) % 60);
-  return `${(min < 10 ? "0" : "") + min}:${secs < 10 ? "0" : ""}${secs}`;
+function convertSecondsInDecimalToHHMMSSFormat(secondsInDecimal) {
```
Comment (amilajack) [inline src/SoundList.js:6]: Any ideas for better names? I recommend taking a look at the date-fns library and seeing what their API naming looks like

**Review by amilajack — CHANGES_REQUESTED** [review]

File: src/SoundList.js (hunk 2 above)
Comment (hamirmahal) [reply src/SoundList.js]: I changed it to `formatTime`.

File: src/SoundList.js [hunk 3]
```diff
@@ -57,7 +62,7 @@ const SoundList = ({
                 </span>
               </div>
               <div
-                className={`group-hover:opacity-100 sm:hover:opacity-0 w-6/12 md:w-4/12 md:space-2 hidden sm:flex flex-row justify-end items-center ${
+                className={`group-hover:opacity-100 sm:hover:opacity-0 w-6/12 md:w-5/12 md:space-2 hidden sm:flex flex-row justify-end items-center ${
```
Comment (hamirmahal) [inline src/SoundList.js:65]: We need to change `md:w-4/12` to `md:w-5/12` to give the sound duration more room since they're longer, now.

File: src/components/Dropdown.js [hunk 4]
```diff
@@ -41,12 +41,12 @@ export default function Dropdown({ onClick = () => {}, disabled = false }) {
...
                     >
                       <a className="flex cursor-pointer justify-between w-full px-4 py-2 text-sm leading-5 text-left">
-                        Export MP3
+                        Download MP3
                       </a>
                     </Menu.Item>
                     <Menu.Item>
                       <a className="flex cursor-pointer justify-between w-full px-4 py-2 text-sm leading-5 text-left">
-                        Export OGG
+                        Download WAV
```
Comment (hamirmahal) [inline src/components/Dropdown.js:49]: I think I rebased things incorrectly, since we're seeing commits originally not in this branch.

**Review by hamirmahal — COMMENTED** [review]
I added a couple of comments that are hopefully explanatory.

File: src/components/Dropdown.js (hunk 4 above)
Comment (amilajack) [reply src/components/Dropdown.js:49]: yea our commit history would be harder to read but it shouldn't break anything. rebasing is tricky!

File: src/SoundList.js [hunk 5]
```diff
@@ -4,10 +4,15 @@ import { playCircleSharp, timeSharp, downloadSharp } from "ionicons/icons";
 import { IonIcon } from "@ionic/react";
 import PlayButton from "./components/PlayButton";
 
-function timeConvert(minutes) {
-  const min = Math.floor(Math.abs(minutes));
-  const secs = Math.floor((Math.abs(minutes) * 60) % 60);
-  return `${(min < 10 ? "0" : "") + min}:${secs < 10 ? "0" : ""}${secs}`;
+function formatTime(secs) {
```
Comment (amilajack) [inline src/SoundList.js:7]: This is a great opportunity to write some unit tests by the way! Here's some boilerplate to help you get started:

```js
// src/__tests__/time.spec.js

import { formatTime } from './path/to/SoundList';

describe('format time', () => {
  it('should format time', () => {
    expect(formatTime(60)).toEqual(someVal)
  });
});
```

Our tests use [jest](jestjs.io), which is a testing library that can write unit tests for react and vanilla js! Think if it as JS's equivalent of JUnit.

File: src/components/Dropdown.js (hunk 4 above)
Comment (amilajack) [reply src/components/Dropdown.js:49]: this is fine tho, don't bother fixing this

File: src/SoundList.js (hunk 5 above)
Comment (amilajack) [reply src/SoundList.js:7]: I think we should add a few unit tests for any edge cases you can identify as well as common cases.

**Review by amilajack — CHANGES_REQUESTED** [review]

File: src/components/Dropdown.js (hunk 4 above)
Comment (hamirmahal) [reply src/components/Dropdown.js]: I think I fixed it with `git reset`, then `git rebase master`, and then `git push -f`. I checked to make sure `git log` returned a history that made sense, at least to me, before doing this.

It seems to have worked, since this pull request is now only showing commits from itself.

I had the option to do `git push -f` earlier, but I tried rebasing use `git pull` this time to get exposure with it, since I've used `git push -f` to fix potential merge conflicts before, and since I wanted to avoid use of the `-f` flag, but it ended up with a messier history.

File: src/__tests__/time.spec.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+    expect(formatTime(3601)).toEqual("01:00:01");
+    expect(formatTime(3600)).toEqual("01:00:00");
+    expect(formatTime(3599)).toEqual("00:59:59");
+    expect(formatTime(601)).toEqual("00:10:01");
+    expect(formatTime(600)).toEqual("00:10:00");
+    expect(formatTime(599)).toEqual("00:09:59");
+    expect(formatTime(61)).toEqual("00:01:01");
+    expect(formatTime(60)).toEqual("00:01:00");
+    expect(formatTime(59)).toEqual("00:00:59");
+    expect(formatTime(0)).toEqual("00:00:00");
```
Comment (amilajack) [inline src/__tests__/time.spec.js:19]: Most sounds don't use hrs formatting. Here's the expectation i think we should have for this function:

```js
expect(formatTime(0)).toEqual("00:00"); // if hrs === 0, do not show it
expect(formatTime(40271)).toEqual("11:11:11"); // if hrs > 0, do show it
```

**Review by amilajack — CHANGES_REQUESTED** [review]

File: src/SoundList.js (hunk 5 above)
Comment (hamirmahal) [reply src/SoundList.js]: I added some basic test cases and covered less fringe edge cases, e.g., formatting 3600 seconds, formatting 3601 seconds, and formatting 3599 seconds.

I didn't add coverage for really extreme edge cases, like if `formatTime` receives negative numbers as a parameter.

For `-1`, it returns `"0-1:59:59"`, but `formatTime` receives floors of track durations, which should always be positive, so `formatTime` should only receive floors of positive numbers, making it always receive nonnegative inputs.

I think we would need to add error handling to `formatTime` if we want to handle more extreme edge cases like these. I'm not sure how "bulletproof" we want to make `formatTime`; it also doesn't check to make sure it receives an input that's of type `Number`.

Thank you for the boilerplate for testing! And, the simile regarding JUnit is helpful. Thanks for teaching me something new!

File: src/SoundList.js [hunk 7]
```diff
@@ -12,6 +12,7 @@ export function formatTime(secs) {
   // Pads a zero in front of a number if it's less than 10
   const padLeft = (num) => `${num < 10 ? "0" : ""}${num}`;
   // e.g., 1 => 01, 9 => 09, 10 => 10, 59 => 59
+  if (hrs === 0) return `${padLeft(mins)}:${padLeft(remainingSecs)}`;
```
Comment (amilajack) [inline src/SoundList.js:15]: nit: we could ternary here

**Review by amilajack — CHANGES_REQUESTED** [review]

File: src/__tests__/time.spec.js (hunk 6 above)
Comment (hamirmahal) [reply src/__tests__/time.spec.js]: I was considering that earlier, but I thought it we might want uniform sound formats. I removed the hours' place of sounds' durations if they're less than an hour in length with the latest commit at the time of this writing.

Removing the hours' place of the sound duration creates a somewhat large gap in the sound data icons, so I also gave the icon denoting the number of downloads a bit more space in the latest commit at the time of this writing.

**Review by amilajack — CHANGES_REQUESTED** [review]
Also some pesky merge conflicts snuck in 😢 . two ways to solve this:
1. merge from master (slightly less clean git history but simpler, this is fine with me for now)
2. rebase (leads to cleaner history)

File: src/SoundList.js (hunk 7 above)
Comment (hamirmahal) [reply src/SoundList.js:15]: It actually uses one more line of code to use the ternary operator, and makes it potentially less human readable, at least for new JavaScript developers, but we can still use a ternary if you prefer.

Original:
```
  if (hrs === 0) return `${padLeft(mins)}:${padLeft(remainingSecs)}`;
  return `${padLeft(hrs)}:${padLeft(mins)}:${padLeft(remainingSecs)}`;
```

Ternary after automatic formatting on save:
```
  return hrs === 0
    ? `${padLeft(mins)}:${padLeft(remainingSecs)}`
    : `${padLeft(hrs)}:${padLeft(mins)}:${padLeft(remainingSecs)}`;
```

Comment (amilajack) [reply src/SoundList.js:15]: Latter approach is considered more idiomatic in js than the first because it slightly reduces cognitive overhead

Comment (amilajack) [reply src/SoundList.js:15]: One of the key benefits of ternaries is that it makes it very easy to visually see the differences between the different return values because they're side by side. One of the things I can't stress enough is that counting LOC isn't actually a good way of determining if something is simpler (but this isn't intuitive at first, takes time to see what I mean).

Comment (hamirmahal) [reply src/SoundList.js]: I agree that the ternary makes it easier to visually see the difference in the two potential return Strings.

The commit I just made uses a ternary.

**Review by amilajack — APPROVED** [review]
🥳 🥳 🥳 good stuff! can't for the demo for this 😍

**Comment by hamirmahal** [comment]
Thank you for the help!
